File: file_comparator.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <filesystem>

bool compareFile(const std::string &file1, const std::string &file2, size_t bufferSize = 4096)
{
    // reading files
    std::ifstream f1(file1, std::ios::binary);
    std::ifstream f2(file2, std::ios::binary);

    if (!f1.is_open() || !f2.is_open())
    {
        std::cerr << "Error! could not open files.\n";
        std::exit(1);
    }

    // Comparing file sizes
    if (std::filesystem::file_size(file1) != std::filesystem::file_size(file2))
    {
        return false;
    }

    std::vector<char> buffer1(bufferSize);
    std::vector<char> buffer2(bufferSize);

    while (true)
    {
        f1.read(buffer1.data(), bufferSize);
        f2.read(buffer2.data(), bufferSize);

        std::streamsize bytesRead1 = f1.gcount();
        std::streamsize bytesRead2 = f2.gcount();

        if (bytesRead1 != bytesRead2)
            return false;

        if (bytesRead1 == 0 && bytesRead2 == 0)
            break;

        if (!std::equal(buffer1.begin(), buffer1.begin() + bytesRead1, buffer2.begin()))
            return false;
    }
    return true;
}

int main(int argc, char *argv[])
{
    std::string file1, file2;
    size_t bufferSize = 4096;

    if (argc >= 3)
    {
        // CLI mode
        file1 = argv[1];
        file2 = argv[2];
        if (argc >= 4)
        {
            try
            {
                bufferSize = std::stoul(argv[3]);
            }
            catch (...)
            {
                std::cerr << "Invalid buffer size. Using default 4096 bytes.\n";
            }
        }
    }
    else
    {
        // Interactive Mode
        std::cout << "Enter path for first file: ";
        std::getline(std::cin, file1);

        std::cout << "Enter path for second file: ";
        std::getline(std::cin, file2);

        std::cout << "Enter buffer size in bytes (default 4096): ";

        if (!(std::cin >> bufferSize) || bufferSize == 0)
        {
            std::cerr << "Invalid buffer size. Using default 4096 bytes.\n";
            bufferSize = 4096;
        }
    }

    // starting the time
    auto start = std::chrono::high_resolution_clock::now();

    bool result = compareFile(file1, file2, bufferSize);

    // stopping the time
    auto end = std::chrono::high_resolution_clock::now();

    // time elapsed
    std::chrono::duration<double> elapsed = end - start;

    if (result)
    {
        std::cout << "Files are identical.\n";
    }
    else
        std::cout << "Files are not identical.\n";

    // time taken for comparison
    std::cout << "Comparison took " << elapsed.count() << " seconds.\n";

    return 0;
}

File: FileComparator.java

import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.InputMismatchException;
import java.util.Scanner;

public class FileComparator {
    public static boolean compareFiles(String file1Path, String file2Path, int bufferSize) {
        Path path1 = Paths.get(file1Path);
        Path path2 = Paths.get(file2Path);

        try {
            if (!Files.exists(path1) || !Files.exists(path2)) {
                System.err.println("Error! could not open files.\n");
                System.exit(1);
            }
            if (Files.size(path1) != Files.size(path2)) {
                return false;
            }

            try (FileInputStream f1 = new FileInputStream(file1Path);
                    FileInputStream f2 = new FileInputStream(file2Path)) {

                byte[] buffer1 = new byte[bufferSize];
                byte[] buffer2 = new byte[bufferSize];
                int bytesRead1, bytesRead2;

                while ((bytesRead1 = f1.read(buffer1)) != -1) {
                    bytesRead2 = f2.read(buffer2);

                    if (bytesRead1 != bytesRead2) {
                        return false;
                    }

                    if (!Arrays.equals(Arrays.copyOf(buffer1, bytesRead1), Arrays.copyOf(buffer2, bytesRead2))) {
                        return false;
                    }
                }
            }
        } catch (IOException e) {
            System.err.println("Error reading files: " + e.getMessage());
            return false;
        }

        return true;
    }

    public static void main(String[] args) {
        String file1Path, file2Path;
        int bufferSize = 4096;
        Scanner scanner = new Scanner(System.in);

        if (args.length >= 3) {
            file1Path = args[0];
            file2Path = args[1];
            try {
                bufferSize = Integer.parseInt(args[2]);
            } catch (NumberFormatException e) {
                System.err.println("Invalid buffer size. Using default 4096 bytes.\n");
            }
        } else {
            System.out.print("Enter path for first file: ");
            file1Path = scanner.nextLine();

            System.out.print("Enter path for second file: ");
            file2Path = scanner.nextLine();

            System.out.print("Enter buffer size in bytes (default 4096): ");
            try {
                bufferSize = scanner.nextInt();
                if (bufferSize == 0) {
                    System.err.println("Invalid buffer size. Using default 4096 bytes.\n");
                    bufferSize = 4096;
                }
            } catch (InputMismatchException e) {
                System.err.println("Invalid buffer size. Using default 4096 bytes.\n");
                bufferSize = 4096;
            }
        }

        long startTime = System.nanoTime();
        boolean result = compareFiles(file1Path, file2Path, bufferSize);
        long endTime = System.nanoTime();

        if (result) {
            System.out.println("Files are identical.");
        } else {
            System.out.println("Files are not identical.\n");
        }

        double duration = (endTime - startTime) / 1_000_000_000.0;
        System.out.printf("Comparison took %.4f seconds.%n", duration);
        scanner.close();
    }
}
